nav2_costmap_2d/src/costmap_2d_publisher.cpp

```diff
@@ -107,85 +107,93 @@ void Costmap2DPublisher::onNewSubscription(const ros::SingleSubscriberPublisher&
 } */
 
 // prepare grid_ message for publication.
-void Costmap2DPublisher::prepareGrid()
+std::unique_ptr<nav_msgs::msg::OccupancyGrid> Costmap2DPublisher::prepareGrid()
 {
   std::unique_lock<Costmap2D::mutex_t> lock(*(costmap_->getMutex()));
-  double resolution = costmap_->getResolution();
+  grid_resolution = costmap_->getResolution();
+  grid_width = costmap_->getSizeInCellsX();
+  grid_height = costmap_->getSizeInCellsY();
+
+  auto grid_ = std::make_unique<nav_msgs::msg::OccupancyGrid>();
```

Comment:
You can leave grid_ as a class variable, in fact, I'd prefer it. You just need to `std::make_unique<nav_msgs::msg::OccupancyGrid>();` on the class variable at the start of prepare grid to instantiate it. I'd revert this commit and just make the simpler change here and in publish grid. You don't want to return smart pointer types.